tsdb/tsm1/reader_range_iterator.go

```diff
@@ -0,0 +1,186 @@
+package tsm1
+
+import (
+	"github.com/influxdata/influxdb/tsdb"
+)
+
+// TimeRangeIterator will iterate over the keys of a TSM file, starting at
+// the provided key. It is used to determine if each key has data which exists
+// within a specified time interval.
+type TimeRangeIterator struct {
+	r    *TSMReader
+	iter *TSMIndexIterator
+	tr   TimeRange
+	err  error
+
+	// temporary storage
+	trbuf []TimeRange
+	buf   []byte
+	a     tsdb.TimestampArray
+}
+
+func (b *TimeRangeIterator) Err() error {
+	if b.err != nil {
+		return b.err
+	}
+	return b.iter.Err()
+}
+
+// Next advances the iterator and reports if it is still valid.
+func (b *TimeRangeIterator) Next() bool {
+	if b.Err() != nil {
+		return false
+	}
+
+	return b.iter.Next()
+}
+
+// Seek points the iterator at the smallest key greater than or equal to the
+// given key, returning true if it was an exact match. It returns false for
+// ok if the key does not exist.
+func (b *TimeRangeIterator) Seek(key []byte) (exact, ok bool) {
+	if b.Err() != nil {
+		return false, false
+	}
+
+	return b.iter.Seek(key)
+}
+
+// Key reports the current key.
+func (b *TimeRangeIterator) Key() []byte {
+	return b.iter.Key()
+}
+
+// HasData reports true if the current key has data for the time range.
+func (b *TimeRangeIterator) HasData() bool {
+	if b.Err() != nil {
+		return false
+	}
+
+	e := excludeEntries(b.iter.Entries(), b.tr)
+	if len(e) == 0 {
+		return false
+	}
+
+	b.trbuf = b.r.TombstoneRange(b.iter.Key(), b.trbuf[:0])
+	var ts []TimeRange
+	if len(b.trbuf) > 0 {
+		ts = excludeTimeRanges(b.trbuf, b.tr)
+	}
+
+	if len(ts) == 0 {
+		// no tombstones, fast path will avoid decoding blocks
+		// if queried time interval intersects with one of the entries
+		if intersectsEntry(e, b.tr) {
+			return true
+		}
+
+		for i := range e {
+			_, b.buf, b.err = b.r.ReadBytes(&e[i], b.buf)
+			if b.err != nil {
+				return false
+			}
+
+			b.err = DecodeTimestampArrayBlock(b.buf, &b.a)
+			if b.err != nil {
+				return false
+			}
+
+			if b.a.Contains(b.tr.Min, b.tr.Max) {
+				return true
+			}
+		}
+	} else {
+		for i := range e {
+			_, b.buf, b.err = b.r.ReadBytes(&e[i], b.buf)
+			if b.err != nil {
+				return false
+			}
+
+			b.err = DecodeTimestampArrayBlock(b.buf, &b.a)
+			if b.err != nil {
+				return false
+			}
+
+			// remove tombstoned timestamps
+			for i := range ts {
```

Comment:
In my mind this may be doing more work than it needs to.

Would it not be possible to do the contains and exclude check at the same time? First build a minimum set of tombstone time ranges, by merging all of the `ts` ranges into a minimum set, and then checking within `Contains` if a timestamp is in a tombstoned range?

Reply:
OK, let me investigate that scenario. For reference, I copied this behavior from compactions:

https://github.com/influxdata/influxdb/blob/d73d73c0d4b708462ef58c8851577486aeb3de56/tsdb/tsm1/file_store_array.gen.go#L923-L927

Reply:
Created #13514 to track this discussion